packages/haunted/useQuery.ts

```diff
@@ -0,0 +1,136 @@
+  extends ApolloQueryMixin(HTMLElement)<TData, TVariables> {
+  declare fetchPolicy: FetchPolicy;
+
+  declare error: ApolloError;
+}
+
+class UseQueryHook<TData, TVariables> extends Hook<
+  [query: DocumentNode, options: QueryHookOptions<TData, TVariables>],
+  QueryResult<TData, TVariables>,
+  QueryBase<TData, TVariables>
+> {
+  declare query: DocumentNode;
+
+  declare variables: TVariables;
+
+  declare client: ApolloClient<NormalizedCacheObject>;
+
+  declare options: QueryHookOptions<TData, TVariables>;
+
+  pollingInterval: number;
+
+  constructor(
+    id: number,
+    state: State<QueryBase<TData, TVariables>>,
+    _query: DocumentNode,
+    _options?: QueryHookOptions<TData, TVariables>
+  ) {
+    super(id, state);
+    const { host } = state;
+    this.id = id;
+    this.state = state;
+    this.options = _options ?? {} as QueryHookOptions<TData, TVariables>;
+    this.query = _query ?? this.options.query;
+    this.update = this.update.bind(this);
+
+    this.initHost();
+
+    QueryBase.prototype.connectedCallback.call(host);
+  }
+
+  initHost(): void {
+    const { options, query, state: { host } } = this;
+
+    const update = () => this.update();
+
+    host.addEventListener('apollo-query-result', update);
+    host.addEventListener('apollo-error', update);
+
+    apply(host, QueryBase, 'query');
+
+    Object.assign(host, {
+      ...DEFAULTS,
+      client: options.client ?? window.__APOLLO_CLIENT__ ?? null,
+      networkStatus: NetworkStatus.ready,
+      noAutoSubscribe: options.noAutoSubscribe ?? host.hasAttribute('no-auto-subscribe'),
+      errorPolicy: options.errorPolicy ?? 'none',
+      fetchPolicy: options.fetchPolicy,
+      pollInterval: options.pollInterval,
+      notifyOnNetworkStatusChange: options.notifyOnNetworkStatusChange,
+      context: options.context,
+      partialRefetch: options.partialRefetch,
+      returnPartialData: options.returnPartialData,
+      shouldSubscribe: options.shouldSubscribe ?? host.shouldSubscribe,
+      query,
+      variables: options.variables,
+    });
+  }
+
+  update(_ = this.query, { variables } = this.options): QueryResult<TData, TVariables> {
+    const { client, data, error, loading, networkStatus } = this.state.host;
+
+    const refetch = this.state.host.refetch.bind(this.state.host);
+
+    // TODO: implement??
+    const updateQuery = () => this.state.host.data;
+
+    return {
+      client,
+      data,
+      error,
+      loading,
+      variables,
+      networkStatus,
+      refetch,
+      updateQuery,
+      fetchMore: this.state.host.fetchMore.bind(this.state.host),
+      subscribeToMore: this.state.host.subscribeToMore.bind(this.state.host),
+      called: undefined,
+      startPolling: (ms: number) =>
+        void (this.pollingInterval = window.setInterval(refetch, ms)),
+      stopPolling: () =>
+        clearInterval(this.pollingInterval),
+    };
+  }
+
+  teardown() {
+    this.state.host.disconnectedCallback();
```

Comment:
Is this for testing? Usually you wouldn't call disconnectedCallback in an element.

Reply:
generally speaking, `connectedCallback` does three things:
1. set up a mutation observer
2. initialize `document` and `variables` from DOM content, if applicable
3. fire an `apollo-element-connected` event, which `<apollo-client>` listens for

https://github.com/apollo-elements/apollo-elements/blob/3e546195fcfa2016ad395494fe79491cbdb9ac70/packages/mixins/apollo-element-mixin.ts#L102-L113

and the `disconnectedCallback` does one thing:
1. fire an `apollo-element-disconnected` event, which `<apollo-client>` listens for.

https://github.com/apollo-elements/apollo-elements/blob/3e546195fcfa2016ad395494fe79491cbdb9ac70/packages/mixins/apollo-element-mixin.ts#L118-L122

Ideally, this behaviour will be uniform across all implementations, and defined only once (i.e. in the mixin)

Reply:
Ok, this was just an observation of mine, don't think you're doing anything wrong necessarily.